core/src/main/java/de/javakaffee/web/msm/storage/RedisStorageClient.java

```diff
@@ -0,0 +1,292 @@
+            @Override protected Boolean execute(BinaryJedis jedis) throws Exception {
+                return jedis.del(keyBytes(key)) == 1;
+            }
+        });
+    }
+
+    @Override
+    public void shutdown() {
+        _pool.shutdown();
+    }
+    
+    private static int convertExp(int exp) {
+        if (exp <= 60*60*24*30) // thirty days
+            return exp;
+        else
+            return Math.max(exp - (int)(System.currentTimeMillis() / 1000), 1);
+    }
+    
+    private static byte[] keyBytes(String key) {
+        return key.getBytes(StandardCharsets.UTF_8);
+    }
+    
+    private abstract class RedisCommandCallable<T> implements Callable<T> {
+        @Override
+        public T call() throws Exception {
+            BinaryJedis jedis = null;
+
+            // Borrow an instance from Jedis without checking it for performance reasons and execute the command on it
+            try {
+                jedis = _pool.borrowInstance(false);
+                return execute(jedis);
+            } catch (JedisConnectionException e) {
+                // Connection error occurred with this Jedis connection, so now make sure to get a known-good one
+                // The old connection is not given back to the pool since it is defunct anyway
+                if (_log.isDebugEnabled())
+                    _log.debug("Connection error occurred, discarding Jedis connection: " + e.getMessage());
+
+                jedis = null;
+            } finally {
+                if (jedis != null)
+                    _pool.returnInstance(jedis);
+            }
+            
+            // Try to execute the command again with a known-good instance
+            try {
+                jedis = _pool.borrowInstance(true);
+                return execute(jedis);
+            } finally {
+                if (jedis != null)
+                    _pool.returnInstance(jedis);
+            }
+        }
+        
+        protected abstract T execute(BinaryJedis jedis) throws Exception;
+    }
+
+    private class JedisPool {
+        private List<BinaryJedis> _instances = new ArrayList<BinaryJedis>();
+
+        public BinaryJedis borrowInstance(boolean knownGood) {
+            synchronized (_instances) {
+                if (_instances.isEmpty()) {
+                    if (_log.isDebugEnabled()) {
+                        _log.debug(String.format("Creating new Jedis instance (host=%s, port=%s, ssl=%s)",
+                            _host, _port, _ssl));
+                    }
+
+                    return createJedisInstance();
+                } else {
+                    if (knownGood) {
+                        // Check all existing connections until we find a good one
+                        BinaryJedis jedis;
+                        do {
+                            jedis = _instances.remove(_instances.size() - 1);
+                            try {
+                                jedis.ping();
```

Comment:
Runs in a synchronized block and might have a negative impact on other users (imagine this instance/connection gets just stuck). The synchronization could be removed by replacing the ArrayList with a ConcurrentLinkedQueue, implementation similar to https://github.com/EsotericSoftware/kryo/blob/master/src/com/esotericsoftware/kryo/pool/KryoPoolQueueImpl.java